src/service/xhr-impl.js

```diff
@@ -344,17 +344,30 @@ function isRetriable(status) {
 /**
  * Returns the response if successful or otherwise throws an error.
  * @paran {!FetchResponse} response
- * @return {!FetchResponse}
+ * @return {!Promise<!FetchResponse|!Error>}
+ * @private Visible for testing
  */
-function assertSuccess(response) {
-  if (!(response.status >= 200 && response.status < 300)) {
-    const err = user.createError(`HTTP error ${response.status}`);
-    if (isRetriable(response.status)) {
-      err.retriable = true;
+export function assertSuccess(response) {
+  return new Promise((resolve, reject) => {
+    if (response.status < 200 || response.status >= 300) {
+      const err = user.createError(`HTTP error ${response.status}`);
+      if (isRetriable(response.status)) {
+        err.retriable = true;
+      }
+      if (response.headers.get('Content-Type') == 'application/json') {
+        return response.json().then(json => {
```

Comment:
You can't return a new promise here - you are still in the constructor of the top-level promise. This is most likely not what you want. You probably just need to run promise w/o return values.